modules/thermal_hydraulics/src/base/Simulation.C

```diff
@@ -268,53 +268,215 @@ Simulation::printComponentLoops() const
+}
+
+void
+Simulation::addSimVariable(bool nl,
+                           const VariableName & name,
+                           FEType fe_type,
+                           const std::vector<SubdomainName> & subdomain_names,
+                           Real scaling_factor)
+{
+  checkVariableNameLength(name);
+
+  if (fe_type.family == SCALAR)
+    mooseDeprecated(
+        "The version of Simulation::addSimVariable() with subdomain names can no longer be used "
+        "with scalar variables since scalar variables cannot be block-restricted. Use the version "
+        "of Simulation::addSimVariable() without subdomain names instead.");
+
+  if (_vars.find(name) == _vars.end()) // variable is new
+  {
+    VariableInfo vi;
+    InputParameters & params = vi._params;
+
+    vi._nl = nl;
+    vi._var_type = "MooseVariable";
+    params = _thm_factory.getValidParams(vi._var_type);
+    params.set<std::vector<SubdomainName>>("block") = subdomain_names;
+
+    auto family = AddVariableAction::getNonlinearVariableFamilies();
+    family = Utility::enum_to_string(fe_type.family);
+    params.set<MooseEnum>("family") = family;
+
+    auto order = AddVariableAction::getNonlinearVariableOrders();
+    order = Utility::enum_to_string<Order>(fe_type.order);
+    params.set<MooseEnum>("order") = order;
+
+    if (nl)
+      params.set<std::vector<Real>>("scaling") = {scaling_factor};
+    else if (!MooseUtils::absoluteFuzzyEqual(scaling_factor, 1.0))
+      mooseError("Aux variables cannot be provided a residual scaling factor.");
+
+    _vars[name] = vi;
+  }
+  else // variable was previously added
   {
     VariableInfo & vi = _vars[name];
-    if (vi._type != type)
-      mooseError(
-          "A component is trying to add variable of the same name but with different order/type");
+    InputParameters & params = vi._params;
+
+    if (vi._nl != nl)
+      mooseError("The variable '",
+                 name,
+                 "' has already been added in a different system (nonlinear or aux).");
+
+    if (vi._var_type != "MooseVariable")
+      mooseError("The variable '",
+                 name,
+                 "' has already been added with a different type than 'MooseVariable'.");
+
+    auto family = AddVariableAction::getNonlinearVariableFamilies();
+    family = Utility::enum_to_string(fe_type.family);
+    if (!params.get<MooseEnum>("family").compareCurrent(family))
+      mooseError("The variable '", name, "' has already been added with a different FE family.");
+
+    auto order = AddVariableAction::getNonlinearVariableOrders();
+    order = Utility::enum_to_string<Order>(fe_type.order);
+    if (!params.get<MooseEnum>("order").compareCurrent(order))
+      mooseError("The variable '", name, "' has already been added with a different FE order.");
+
+    // If already block-restricted, extend the block restriction
+    if (params.isParamValid("block"))
+    {
+      auto blocks = params.get<std::vector<SubdomainName>>("block");
```

Comment:
`ANY_BLOCK_ID` should not need consideration. While it's possible for a developer to try and pass this, they should always be passing actual block names corresponding to that component - they should always be block-restricting their variables.

Reply:
so maybe we should assert against it?
lists of blocks arent supposed to be : block_1, block_2, "any_block_id" (the name)

I can see the NS component stuff being set to any block id if it s the only component?

Reply:
> so maybe we should assert against it?
lists of blocks arent supposed to be : block_1, block_2, "any_block_id" (the name)
>
> I can see the NS component stuff being set to any block id if it s the only component?

Just to clarify, users do not have any input regarding the blocks for components. Only the developers do, so this would only be a check for developers. And I do not think we should ever allow the developer to use `ANY_BLOCK_ID` - it just doesn't make sense when you have the requirement that multiple components are allowed.